core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -355,8 +367,28 @@ final class ShuffleBlockFetcherIterator(
                 updateMergedReqsDuration(wasReqForMergedChunks = true)
                 results.put(FallbackOnPushMergedFailureResult(
                   block, address, infoMap(blockId)._1, remainingBlocks.isEmpty))
-              } else {
+              } else if (!shouldPerformShuffleLocationRefresh) {
                 results.put(FailureFetchResult(block, infoMap(blockId)._2, address, e))
+              } else {
+                val (shuffleId, mapId) = BlockId.getShuffleIdAndMapId(block)
```

Comment:
Should we move the `getShuffleIdAndMapId` into the `Try` ?
We will effectively block shuffle indefinitely in case `getShuffleIdAndMapId` throws an exception (it should not currently - but code could evolve).

Something like:
```
                Try {
                  val (shuffleId, mapId) = BlockId.getShuffleIdAndMapId(block)
                  mapOutputTrackerWorker
                    .getMapOutputLocationWithRefresh(shuffleId, mapId, address)
                } match {
```